### PR TITLE
feat: migrate to alloy TxLegacy

### PR DESCRIPTION
**Pull request by leruaa**
Cf #9484

### CODE REVIEW

File: crates/primitives/src/transaction/legacy.rs
Comment (mattsse) [inline crates/primitives/src/transaction/legacy.rs:1]: amazing

File: crates/primitives/src/transaction/mod.rs
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:1180]: maybe worth exploring if we can phase out the reth-prim signature type?

File: crates/primitives/src/transaction/signature.rs
Comment (mattsse) [inline crates/primitives/src/transaction/signature.rs:47]: r,s?

File: crates/storage/codecs/src/alloy/transaction.rs
Comment (mattsse) [inline crates/storage/codecs/src/alloy/transaction.rs:99]: this is the same for both and doesn't require an op feature

File: crates/storage/codecs/src/alloy/transaction.rs
Comment (mattsse) [inline crates/storage/codecs/src/alloy/transaction.rs:1]: with https://github.com/paradigmxyz/reth/pull/9488 this will require a smol change, basicall to_compact now accepts &self so we need a ref<'a> type as well

File: crates/storage/codecs/src/alloy/transaction.rs
Comment (mattsse) [inline crates/storage/codecs/src/alloy/transaction.rs:9]: can remove most of this
```suggestion
```

**Review by mattsse — CHANGES_REQUESTED** [review]
amazing work so far!

**Comment by leruaa** [comment]
@mattsse the test `test_sending_invalid_transactions` is failling because it's creating an invalid signature, but now with the conversion to alloy `Signature` we create a `k256::ecdsa::Signature` and it fails.

Should as_alloy_signature() returns a Result and also update the callers to also returns a Result?

**Comment by mattsse** [comment]
> but now with the conversion to alloy Signature we create a k256::ecdsa::Signature and it fails.

can you elaborate? unclear what fails, the default alloy `Signature` type is feature gated, imo we should use `Signature<()>` here 

> it wouldn't be possible to create invalid signatures anymore.

I think with `Signature<()>` this would still be possible, but perhaps we need to tune something for this type?

**Comment by leruaa** [comment]
@mattsse Unfortunately alloy-primitive "k256" feature is enabled due to this: 

https://github.com/alloy-rs/alloy/blob/423a4c53c1bffdfb2acca2daee60215194f7711d/crates/signer/Cargo.toml#L22C7-L22C18

**Comment by leruaa** [comment]
To add a bit of context:

The issue with the test [test_sending_invalid_transactions](https://github.com/paradigmxyz/reth/blob/4a1916155752db84679315aa56175bf3f331adac/crates/net/network/tests/it/txgossip.rs#L102) is that it create an invalid signature here:

https://github.com/paradigmxyz/reth/blob/4a1916155752db84679315aa56175bf3f331adac/crates/net/network/tests/it/txgossip.rs#L132

Deeper in the code this signature is converted to an alloy `Signature`, but as the `k256` feature is on, a `k256::ecdsa::Signature` is needed but can't be created from an invalid signature.

I fixed `test_sending_invalid_transactions()` by creating a valid signature, but more generally I guess we should be able to convert reth signatures to alloy ones even if they are invalid.

I think the main issue here is that the `k256` feature on alloy-primitives is not additive (i.e. it remove the `Signature<()>` type when activated), and that is an issue in reth because we won't be able to make it off (it's needed in revm).

So I think we need to update alloy-primitives to never remove `Signature<()>` and rename `Signature` when `k256` feature is enabled to something else.

@mattsse wdyt? I hope to have been clear :)

**Comment by mattsse** [comment]
> i.e. it remove the Signature<()> type when activated), and that is an issue in reth because we won't be able to make it off (it's needed in revm).

yeah, tbh I find this behaviour in alloy-primitives a bit strange because this obfuscates things.
I'd like if we could add a simple alias for `Signature<()>` to alloy-primitives, like `RawSignature` or something,

but maybe you can use a type alias for this here for now?

we also need to change these functions

https://github.com/alloy-rs/alloy/blob/main/crates/consensus/src/transaction/legacy.rs#L120-L120

so they accept any Signature<S> because we only need the rsv fields

**Comment by leruaa** [comment]
Unfortunately we can't use a type alias because `Signature<S>` is private.

**Comment by mattsse** [comment]
this is kinda stupid lol

**Comment by mattsse** [comment]
@leruaa just checking in, is this now only blocked by a new alloy release?

**Comment by leruaa** [comment]
Yes exactly

**Comment by mattsse** [comment]
this should be unblocked now, main is on newest alloy now

**Comment by leruaa** [comment]
@mattsse Unfortunately there is an unexpected complication, initially a test was in error because the chain_id was lost during a `TransactionSigned ` encode/decode, so I changed `odd_y_parity` bool to `parity` Parity type, but now other tests are failing because derive arbitrary impl for Parity can produce invalid Parity::Eip155(0) that is decoded to Parity::Parity(false), causing the assert to fail.

File: crates/primitives/src/transaction/mod.rs
Comment (leruaa) [reply crates/primitives/src/transaction/mod.rs]: That would be interesting, but can it be done in a followup?

**Review by emhane — APPROVED** [review]
nice that you left the TODOs for replacing this temp types that are needed to be able to impl transition to alloy in many PRs and not one huge one.

**Review by mattsse — COMMENTED** [review]
this is a pretty huge unlock and should unblock further alloy changes.

ptal at the codec things @joshieDo

**Review by joshieDo — APPROVED** [review]
LGTM, tried locally with no issues

**Comment by mattsse** [comment]
sweet, blocking this to go in after tmrw's release

File: crates/primitives/src/transaction/mod.rs
Comment (emhane) [inline crates/primitives/src/transaction/mod.rs:1445]: revert these unrelated semicolon additions pls

**Review by emhane — CHANGES_REQUESTED** [review]

File: crates/primitives/src/transaction/mod.rs
Comment (emhane) [reply crates/primitives/src/transaction/mod.rs:1445]: reth uses `cargo +nightly fmt`, try the `make lint` command

**Review by emhane — APPROVED** [review]

File: crates/primitives/src/transaction/mod.rs
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:1033]: this is only used for testing, right?

could we move this into the test module?

File: crates/primitives/src/transaction/signature.rs
Comment (mattsse) [inline crates/primitives/src/transaction/signature.rs:239]: why was this test removed?

File: crates/primitives/src/transaction/signature.rs
Comment (mattsse) [inline crates/primitives/src/transaction/signature.rs:254]: same

**Review by mattsse — CHANGES_REQUESTED** [review]
alright, let's finally do this.

tysm for kickstarting this.

only have two more qs

File: crates/primitives/src/transaction/signature.rs
Comment (leruaa) [reply crates/primitives/src/transaction/signature.rs:239]: The corresponding method payload_len_with_eip155_chain_id() has been removed.

File: crates/primitives/src/transaction/signature.rs
Comment (leruaa) [reply crates/primitives/src/transaction/signature.rs:254]: encode_with_eip155_chain_id() has also been removed

**Review by mattsse — COMMENTED** [review]
sorry I have to ask more questions because I'm a bit lost with the signature types

also would appreciate a closer look by @Rjected

File: crates/primitives/src/transaction/signature.rs
Comment (leruaa) [reply crates/primitives/src/transaction/signature.rs:248]: This was to reflect the "else" case here:
https://github.com/paradigmxyz/reth/blob/ae8ce20c0b433f401f596fb7860b4e66a8657a61/crates/primitives/src/transaction/signature.rs#L88-L103

**Comment by leruaa** [comment]
I realize this is not taken into account anymore:

https://github.com/paradigmxyz/reth/blob/ae8ce20c0b433f401f596fb7860b4e66a8657a61/crates/primitives/src/transaction/signature.rs#L93-L100

Let me add a commit

**Review by mattsse — APPROVED** [review]
lgtm, let's send it this finally @klkvr

**Pull request merged**
